### DISCUSSION TITLE
Releases?

### DISCUSSION THREAD

**Discussion by LDprg** [winapps-org/winapps#23] [Ideas]
How are we doing on release, because most distro package manager require them.

There are view things we should think about:
- Release Versioning (0.1 or 0.0.1 or sth. else)
- Release Management, are we going to use the default GitHub releases or do we create a release branch
- How are we dealing with the rewrite and legacy packages so there is no collision (design them to work in parallel or only allow one or the other) 
- Distribution, which package managers and how their requirements are
- For the rewrite we should consider uploading the code to [crates.io](https://crates.io), which would allow installation via cargo

**Comment by LDprg** [comment]
As I am using Arch (btw) I would take care about the AUR package. As it is easier to get a package there than in the pacman repos.

**Comment by oskardotglobal** [comment]
For the versioning I'd go with good old semantic versioning: https://semver.org/

Github releases would also be a good idea. We could just start by distributing binaries as tarballs via github releases and then go from there. As for deb-based and rpm-based distros (I personally use fedora), downloads with the deb and the rpm will suffice (On fedora you can directly install software via a link to the rpm using `dnf`) 

Also, for the collision problem I'd just name legacy "winapps" and then rewrite would be "winapps2"
And at some point we'd just abandon winapps2 and upload the rewrite artifactes under "winapps"

**Reply by LDprg** [reply]
It is probaply the best to distribute deb and rpm for now. Maybe we could consider the [Debian community repository](https://mpr.makedeb.org/ ) and rpm fusion for later.

**Reply by LDprg** [reply]
What if we use winapps-legacy and winapps or the other way around winapps and winapps-rust or something like that for naming.

This would probably cause less confusion.

**Reply by AkechiShiro** [reply]
I would go for your choice @LDprg but most likely not renaming the old binary as winapps-legacy but leaving it as is for (old users/not creating confusion for old users/conflicts in pkg managers), and I would go for : 
- Legacy : `winapps`
- New : `winapps-rs` (shorter than `winapps-rust`)
- New proposal 2 : `winapps-ng`

**Comment by LDprg** [comment]
Also how about AppImages or Flatpacks for non Debian Fedora Arch based distros?

**Reply by fenglengshun** [reply]
Speaking personally, Flatpak would be great for most distro, especially for Steam Deck users. The problem would be working with the build process, what to ship, and how to properly interface with the VM. I'd recommend just going with a .flatpak release on github first, like Rustdesk, as I'd imagine adding Flathub's restrictions is going to be difficult even with flatpak-spawn now being more accepted.

AppImage... I would put it as secondary - there's a lot more Flatpak users due to Steam Deck, Fedora, and non-Snap Ubuntu-based. I don't use them enough to really say, though I don't really like how they don't have a package manager.

Outside of that, there's also the option of Nix which is also distro-agnostic.

Snaps not under consideration?

**Reply by oskardotglobal** [reply]
This is a great idea. Lets target Flatpaks first. If we have Flatpaks, there's also no reason to target either AppImage or snap (but honestly, I'd prefer to not target snaps at all even if there were reasons to do so)

**Reply by fenglengshun** [reply]
If you're targeting Flathub, maybe you can use Gnome Boxes as dependency? Though that would mean the rest of the project would need to be adjusted with assumption of gnome-boxes on flatpak as one of the VM runner.

**Reply by oskardotglobal** [reply]
We won't use gnome boxes, we'll use https://github.com/quickemu-project/quickemu
Also see #8

**Reply by fenglengshun** [reply]
Then you'll probably want to look into `flatpak-spawn`. That seems to be what Lutris used to get around some of the Flatpak limitations.

**Reply by LDprg** [reply]
I personally would also prefer to not providing snap packs (from a technical and ideological perspective).

Maybe we should focus on Flatpacks first since winapps itself hast big performance requirements and Flatpacks are cross platform.

So I would do like:
1) Tarballs
2) Flatpacks
3) Aur, rpm, deb
4) AppImages (optionally, here we could also wait for user made ones)

**Reply by freechelmi** [reply]
Flatpak is a great way to target all users at once , even chromeOS users.

I'm not experience in flatpak but I can have a look as I was willing to learn it anyway

**Reply by fenglengshun** [reply]
AUR, someone will probably make them. I think it's best to just allow the AUR community to make and maintain the PKGBUILD, and only monitor and give advice only when there's an issue. Unless someone in the new winapps teams wants to maintain it themselves?

Deb and RPM would be great. Maybe use OBS to build the binary then?

**Reply by fbartels** [reply]
In the past we had an obs installation at work to create packages through it. even when using their public instance I don't think its worth the effort for a single package. Better to use a github action to create these.

**Reply by fenglengshun** [reply]
Fair enough. Should probably have an update-check mechanism then - one advantage of OBS I recall was that you can add them as repo on both Debian and Ubuntu (depending on how it's setup) which is nice for updating.

**Reply by PaulCombal** [reply]
As an Ubuntu user, Snap and Flatpak make the most sense. Debian packages would typically be used for system dependencies only. Correct me if I'm wrong.

**Comment by oskardotglobal** [comment]
What about nix? We could also use that, quickemu uses it.

**Reply by LDprg** [reply]
I not very familiar with nix, but is it compatible with flatpacks?

**Reply by oskardotglobal** [reply]
Nix is a package manager (and there's also a distro using it) which allows to you write state as code a bit like ansible. It would basically be a script very similar to the AURs PKGBUILD.
However, nix is cross-platform and I'm pretty sure we could depend on quickemu which would make the whole installation a lot easier.

**Reply by LDprg** [reply]
Seems perfect for winapps and nix package manager seems quite common for distribution.

**Reply by fenglengshun** [reply]
Nix will be tough, I don't recommend it to start with. You'll need to learn a lot about how Nix works, and have to pass a pretty strict test to ensure it is reliable and reproducible.

The Nix userbase is growing (even SteamOS including an offload for `/nix` now in Main branch) but Flatpak is still the much more popular solution due to its simplicity with PackageKit allowing for easy GUI installation.

**Reply by oskardotglobal** [reply]
I know someone who might be able to help with that

**Reply by AkechiShiro** [reply]
@bayazidbh I'm maintaining currently 2 pkgs in Nixpkgs in Nix, Azure-cli and Bloodhound, I'm still a newbie at writing a NixOS module but I believe we can find a way to move forward using Nix.
Quickemu is already packaged in Nix as well : https://github.com/NixOS/nixpkgs/blob/nixos-23.05/pkgs/development/quickemu/default.nix#L93

Hence giving us a immense headstart, we'll depend on it, from what I've read so far.

**Reply by AkechiShiro** [reply]
Also this could help us in packaging Rust with Nix : https://www.youtube.com/watch?v=HGIKY2XpCFU
I was at Nixcon 2022, and I met the author of this talk.
